### PR TITLE
Fix the smoke animation when demolishing a castle building (Battlefield).

### LINKED ISSUE #3242: Combat, animation of the drawbridge destruction
Currently in fheroes2, when the drawbridge is destroyed, the part above the drawbridge is removed after the dust cloud animation.
It would be more aesthetic to remove the part above the castle gate at the same time as the dust cloud animation is displayed (to hide the removal of the part) and not after.

In fheroes2:

https://user-images.githubusercontent.com/43583994/115622977-de4b5480-a2f8-11eb-8676-f15f01e98bf8.mp4

Wrathmont attacks Dragontooth.

Here is the save file:
[Before attack defense freeze.zip](https://github.com/ihhub/fheroes2/files/6353760/Before.attack.defense.freeze.zip)

### LINKED ISSUE #6222: Battlefield, castle assault, after boulder impact, the building is demolished only after the smoke animation
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

In the castle assault, if catapult attack is successful, after the impact of the boulder, the building is demolished only after the animation of the smoke. And it should be demolished in the middle of the smoke animation.
fheroes2:

https://user-images.githubusercontent.com/113276641/202837831-3e0655cb-8584-47fc-b1e0-53be3d030ba3.mp4

original game:

https://user-images.githubusercontent.com/113276641/202475809-2d5fd120-5693-48e3-8517-6a58399a78b1.mp4

### Save file

[0_9962a_0001.zip](https://github.com/ihhub/fheroes2/files/10048937/0_9962a_0001.zip)

### Additional info

The function that removes the building when it is hit (`SetCastleTargetValue`) is called only after the full animation of the catapult attack (`RedrawActionCatapult`).
The possible solution may be to separate smoke animation into separate two functions: the smoke animation before demolition of the building and after it. (`SetCastleTargetValue` cannot be called from `RedrawActionCatapult` as it is a private function of Arena class)

### PR DESCRIPTION
**Pull request by Districh-ru**
close #6222 
close #3242

https://user-images.githubusercontent.com/113276641/205485764-20edabf9-387c-4e29-a52d-fd9394fa8dbd.mp4

### CODE REVIEW

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @Districh-ru , I left multiple comments in this pull request. Could you please take a look at them when you have time?

File: src/fheroes2/battle/battle_action.cpp [hunk 1]
```diff
@@ -1000,6 +1000,10 @@ void Battle::Arena::ApplyActionCatapult( Command & cmd )
 
                 if ( hit ) {
                     SetCastleTargetValue( target, GetCastleTargetValue( target ) - damage );
+                    if ( _interface ) {
+                        // Continue animating the smoke cloud after changing the "health" of the building.
+                        _interface->RedrawAfterDemolitionSmoke( target );
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/battle/battle_action.cpp:1005]: Hi @ihhub, I agree.
However, I suggest renaming `RedrawActionCatapult` to `RedrawActionCatapultPart1` and the introduced method to `RedrawActionCatapultPart2` to match the names of other methods.

File: src/fheroes2/battle/battle_arena.cpp [hunk 2]
```diff
@@ -1075,10 +1075,6 @@ void Battle::Arena::SetCastleTargetValue( int target, uint32_t value )
     case CAT_BRIDGE:
         if ( _bridge->isValid() ) {
             if ( !_bridge->isDown() ) {
-                if ( _interface ) {
-                    _interface->RedrawBridgeAnimation( true );
-                }
-
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/battle/battle_arena.cpp:1081]: This change was made to match the original game: after the boulder hits, the bridge lift breaks and the bridge falls and collapses - that's why we have a second smoke cloud.

File: src/fheroes2/battle/battle_interface.cpp [hunk 3]
```diff
@@ -4194,6 +4206,58 @@ void Battle::Interface::RedrawActionCatapult( int target, bool hit )
...
+    if ( isBridgeDestroyed ) {
+        pt2.x = pt1.x - 45;
+        pt2.y = pt1.y + 65;
+        // Increase maxAnimationFrame to finish the second smoke animation.
+        maxAnimationFrame += 2;
+        // Bridge smoke animation should contionue from the 7th frame.
+        frame = 6;
+    }
+
+    while ( le.HandleEvents() && frame < maxAnimationFrame ) {
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/battle/battle_interface.cpp:4241]: If we miss this method is not called:
```cpp
                if ( hit ) {
                    SetCastleTargetValue( target, GetCastleTargetValue( target ) - damage );
                    if ( _interface ) {
                        // Continue animating the smoke cloud after changing the "health" of the building.
                        _interface->RedrawActionCatapultPart2( target );
                    }
                }
```

File: src/fheroes2/battle/battle_arena.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/battle/battle_arena.cpp:1081]: Original game:

https://user-images.githubusercontent.com/113276641/205684780-178f2e34-6081-4ba2-aaba-8b622990bb6a.mp4

[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/battle/battle_arena.cpp:1081]: And if the bridge is alerady open, we also have a second smoke cloud.
Maybe the bridge mechanism and the wall fragments fall on the bridge and break it. (while the creature on the bridge somehow manages to dodge)
Original game:

https://user-images.githubusercontent.com/113276641/205686212-ad32fe3c-b1a5-48c8-9254-1b79e6fc97a1.mp4

File: src/fheroes2/battle/battle_action.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/battle/battle_action.cpp]: Looks like to me. Not the greatest but it's better than now.

File: src/fheroes2/battle/battle_interface.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/battle/battle_interface.cpp:4247]: Ah, got it.

File: src/fheroes2/battle/battle_interface.cpp [hunk 4]
```diff
@@ -4194,6 +4215,55 @@ void Battle::Interface::RedrawActionCatapult( int target, bool hit )
...
+    if ( !isHit ) {
+        catapult_frame = 0;
+    }
+}
+
+void Battle::Interface::RedrawActionCatapultPart2( const int catapultTargetId )
+{
+    // Finish the smoke cloud animation after the building's state has changed after the hit and it is drawed as demolished.
+
+    LocalEvent & le = LocalEvent::Get();
```
Comment (ihhub) [inline src/fheroes2/battle/battle_interface.cpp:4227]: A usual practice is to declare variables just before their usage (ignore the most of the code in this file as it was mostly written before us). I suggest to put this variable at line 4246 just before the loop.

**Review by ihhub — APPROVED** [review]
Hi @Districh-ru , I essentially approve this pull request. I left one optional code style comment but it's up to you to change it.

**Comment by ihhub** [comment]
Hi @Districh-ru , based on your great work in animations during battle I recommend to try your skills in fixing these issues: #5701, #2314, #1229 and #1041. 🙂

File: src/fheroes2/battle/battle_interface.cpp (hunk 4 above)
Comment (ihhub) [reply src/fheroes2/battle/battle_interface.cpp]: You're welcome!

**Review by Branikolog — APPROVED** [review]
Well done!

**Comment by ihhub** [comment]
@Districh-ru , a bunch of thanks for this improvement!

**Comment by Districh-ru** [comment]
> Hi @Districh-ru , based on your great work in animations during battle I recommend to try your skills in fixing these issues: #5701, #2314, #1229 and #1041. 🙂

Hi @ihhub, thanks ;)
I was already thinking about #5701

**Pull request merged**